jetty-http2/http2-hpack/src/main/java/org/eclipse/jetty/http2/hpack/HpackEncoder.java

```diff
@@ -180,26 +176,30 @@ else if (metadata.isResponse())
             encode(buffer, status);
         }
 
-        // Add all non-connection fields.
+        // Remove fields as specified in RFC 7540, 8.1.2.2.
         HttpFields fields = metadata.getFields();
         if (fields != null)
         {
-            Set<String> hopHeaders = fields.getCSV(HttpHeader.CONNECTION, false).stream()
-                .filter(v -> specialHopHeaders.get(v) == Boolean.TRUE)
-                .map(StringUtil::asciiToLowerCase)
-                .collect(Collectors.toSet());
+            // For example: Connection: Close, TE, Upgrade, Custom.
+            List<String> values = fields.getCSV(HttpHeader.CONNECTION, false);
+            Set<String> hopHeaders = new HashSet<>();
+            for (String value : values)
+            {
+                // Keep TE as a special case.
+                if (!"TE".equalsIgnoreCase(value))
+                    hopHeaders.add(StringUtil.asciiToLowerCase(value));
+            }
             for (HttpField field : fields)
             {
-                if (field.getHeader() == HttpHeader.CONNECTION)
+                HttpHeader header = field.getHeader();
+                if (header != null && HEADERS_TO_REMOVE.contains(header))
                     continue;
-                if (!hopHeaders.isEmpty() && hopHeaders.contains(StringUtil.asciiToLowerCase(field.getName())))
+                if (hopHeaders.contains(StringUtil.asciiToLowerCase(field.getName())))
                     continue;
-                if (field.getHeader() == HttpHeader.TE)
+                if (header == HttpHeader.TE)
                 {
-                    if (!field.contains("trailers"))
+                    if (!"trailers".equalsIgnoreCase(field.getValue()))
```

Comment:
This is a change in behavior. Previously we ignored all elements of a `TE` header other than `trailers`, now if a `TE` header is not exactly "trailers" it is not copied.

Reply:
Also, CONNECTION is an ignored header, so I think you need to add CONNECTION_TE if you have trailers.  or do we????? - Reading [the spec](https://http2.github.io/http2-spec/#rfc.section.8.1.2.2) it is not entirely clear, but I think you are probably right that we should just add the TE header without the connection header.

Reply:
Also Also, TE value can be a quality lists, so "trailers; q=1.0" is possible, so using HttpField.contains will handle that